Use jsdom polyfill and update jest

NSeydoux opened the pull request:
Initially, the intent of this PR was just to use jsdom polyfills we provide to simplify the environment setup. However, it showed that we are currently not running our unit test using the browser jose bundle in the browser package, because we are using an older version of `jest`. So this PR also looks at updating jest to the latest version.

EDIT: a bunch of upgrades tacked on this PR: 
- Update to the latest linting rules (because of the polyfill update coming from the same repo)
- Migrated jest configs to TS to reuse explicit typing provided by `ts-jest`
- A bunch of tests required some changes because of the jest major update
- The polyfill still don't entirely cut it, so for now the custom environment is still used. Changes will be pushed to the polyfills, and when upgrading the custom environment will be removed.

NSeydoux committed:
Remove InruptError class

The InruptError class is making use of the inrupt vocab library, which relies on some local storage behavior. This behavior breaks under jest 29, and since thiese features were not used in this library and should be implemented in a proper error library instead, the dependency is removed altogether moving to jest 29.

NSeydoux committed:
Prepare core package for jest 29

Fix incorrect tests

NSeydoux committed:
Use inrupt polyfills in jest setup

NSeydoux committed:
Update jest version to 29

Install latest jose as well

NSeydoux committed:
Fix test in oidc package for jest 29 migration

NSeydoux committed:
WIP Try supporting ESM with Jest 29

Jest 29 has multiple improvements, and in particular it adds support for the "browser" field in the package.json if using jsdom. However, jose ships as ESM-only for the browser module, so we must support ESM there. In any case, this means that we were not testing with the same code that gets deployed to our customers, which is a problem we must resolve no matter what.

Currently, this doesn't work, importing jose fails when running the oidc package test (the first package in the browser dependency chain).

ThisIsMissEm commented on jest.config.js:
```diff
@@ -2,9 +2,8 @@ module.exports = {
-    "^.+\\.(ts|tsx)?$": "ts-jest",
-  },
+  preset: "ts-jest/presets/default-esm",
+  extensionsToTreatAsEsm: ['.ts'],
```
Isn't this default?

NSeydoux replied:
The default is `ts-jest`, which outputs CJS modules, while `ts-jest/presets/default-esm` outputs ESM.

ThisIsMissEm commented on package.json:
```diff
@@ -42,17 +43,18 @@
     "eslint-plugin-import": "^2.22.1",
-    "eslint-plugin-jest": "^26.0.0",
+    "eslint-plugin-jest": "^27.1.1",
     "eslint-plugin-prettier": "^4.0.0",
```
You can remove most, if not all of these, they now come from the shared config package.

NSeydoux replied:
Done in 8e4102341a4f85e56e788421c4e7d5491754bb78 (also wow a lot of linting things overdue)

ThisIsMissEm commented on packages/browser/jest.config.js:
```diff
@@ -1,7 +1,8 @@
   ...require("../../jest.config"),
   // Allows running tests from packages/browser/
-  testEnvironment: "<rootDir>/../../tests/environment/customEnvironment.js",
+  setupFilesAfterEnv: ['<rootDir>/../../jest.setup.js'],
```
You shouldn't need this again as it's in the root config.

NSeydoux replied:
`rootDir` doesn't resolve to the same directory whether the `jest` command runs from the root or from a sub-package of the monorepo. Overriding this here allows to run `npm run test` from `packages/browser/`, which is useful when you don't want to run the whole test suite. If this isn't overriden there, `setupFilesAfterEnv` points to `packages/browser/jest.setup.js`.

ThisIsMissEm replied:
it'd be better to set `rootDir` option, I think, and specify that as `rootDir: "../../"` I think?

NSeydoux replied:
Doing so runs all the test suites, and not only the target one, so it doesn't quite solve the issue.

ThisIsMissEm replied:
Ah, alrighty!

ThisIsMissEm commented on packages/oidc/jest.config.js:
```diff
@@ -1,5 +1,6 @@
 module.exports = {
   ...require("../../jest.config"),
+  testEnvironment: "jsdom",
```
Might be worth leaving a comment that this package is only used by browser code so we only test in jsdom

ThisIsMissEm commented on packages/oidc/package.json:
```diff
@@ -20,7 +20,7 @@
     "format": "prettier --write \"src/**\" \"__tests__/**\"",
     "licenses:check": "license-checker --production --out license.csv --failOn \"AGPL-1.0-only; AGPL-1.0-or-later; AGPL-3.0-only; AGPL-3.0-or-later; Beerware; CC-BY-NC-1.0; CC-BY-NC-2.0; CC-BY-NC-2.5; CC-BY-NC-3.0; CC-BY-NC-4.0; CC-BY-NC-ND-1.0; CC-BY-NC-ND-2.0; CC-BY-NC-ND-2.5; CC-BY-NC-ND-3.0; CC-BY-NC-ND-4.0; CC-BY-NC-SA-1.0; CC-BY-NC-SA-2.0; CC-BY-NC-SA-2.5; CC-BY-NC-SA-3.0; CC-BY-NC-SA-4.0;  ...
-    "test": "jest --coverage --verbose",
+    "test": "NODE_OPTIONS=--experimental-vm-modules jest --coverage --verbose",
```
Be sure to check which node version introduced this

NSeydoux replied:
No longer required

ThisIsMissEm reviewed (commented):
Looks good overall, a few minor comments

NSeydoux committed:
Setup jest to transform ESM deps to CJS

Overriding the transformIgnorePatterns forces jest to transform the content of the node_module, which, combined to the ts-jest preset which transforms both TS and JS and to the allowJs config set to true in the tsconfig allows jest to run tests on ES modules by transforming them to CJS.

NSeydoux committed:
Make jest configs TS

This allows to treat them as modules, as well as using types for typing hints.

NSeydoux committed:
Use rollup to output ESM from core module

NSeydoux committed:
Update to the latest jose

NSeydoux committed:
Polyfill jsdom crypto

NSeydoux committed:
Fix tests

NSeydoux committed:
Update browser module to jest 29

NSeydoux committed:
Migrate node module to jest 29

NSeydoux committed:
fixup! Make jest configs TS

NSeydoux committed:
Upgrade to latest linter rules

NSeydoux committed:
fixup! fixup! Make jest configs TS

NSeydoux committed:
Change import for node e2e test setup

NSeydoux committed:
fixup! Upgrade to latest linter rules

ThisIsMissEm commented on jest.setup.ts:
```diff
@@ -0,0 +1 @@
+import "@inrupt/jest-jsdom-polyfills";
```
```suggestion
import "@inrupt/jest-jsdom-polyfills";

```

missing trailing new line, and eslint may complain about license header.

ThisIsMissEm commented on jest.config.ts:
```diff
@@ -1,10 +1,11 @@
-    "^.+\\.(ts|tsx)?$": "ts-jest",
-  },
+  preset: "ts-jest/presets/js-with-ts",
+  transformIgnorePatterns: [],
```
let's add a comment as to why.

NSeydoux replied:
53b454b1bf7188d14e2e793db709c96a74a37cd4

ThisIsMissEm commented on package.json:
```diff
@@ -30,30 +30,28 @@
+    "@inrupt/eslint-config-base": "^1.3.0",
+    "@inrupt/eslint-config-lib": "^1.3.0",
+    "@inrupt/jest-jsdom-polyfills": "^1.3.0",
+    "@peculiar/webcrypto": "^1.4.0",
```
This should be done in `@inrupt/jest-jsdom-polyfills`

NSeydoux replied:
Waiting on https://github.com/inrupt/javascript-style-configs/pull/62

ThisIsMissEm replied:
Approved

NSeydoux replied:
Now for the release :) https://github.com/inrupt/javascript-style-configs/pull/63

ThisIsMissEm commented on tests/environment/customEnvironment.ts:
```diff
@@ -0,0 +1,25 @@
+    (this.global.crypto.subtle as any) = (new Crypto()).subtle;
+    this.global.CryptoKey = CryptoKey;
+  }
+};
```
Let's fix this in jest-jsdom-polyfills instead of using custom environments.

NSeydoux replied:
This works in the polyfill (I have tested it locally, I was planning on merging this and revisiting after releasing the updated polyfills), but I couldn't get the `Uint8Array` thing to work in there, so for now I'm afraid the custom environment has to stay. We can revisit why, but I'd like to move on because this has been a time sink :)

NSeydoux replied:
See https://github.com/inrupt/javascript-style-configs/pull/62

ThisIsMissEm replied:
I think this should've been released now, so this shouldn't be necessary anymore.

ThisIsMissEm commented on tsconfig.build.json:
```diff
@@ -14,7 +14,9 @@
     "sourceMap": true,
     "lib": ["es2018", "dom"],
     "moduleResolution": "node",
```
might be worth mentioning this kinda changes in a way soon-ish: https://www.typescriptlang.org/tsconfig#moduleResolution

ThisIsMissEm reviewed (changes requested):
Let's get the changes needed into jest-jsdom-polyfills instead of using a custom environment here.

NSeydoux committed:
Move crypto polyfilling to the polyfill module

NSeydoux committed:
Add comment to jest config

NSeydoux committed:
Update jest.setup.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Use subpatch imports

NSeydoux committed:
Fix subpath imports

NSeydoux committed:
Skip tests with open handles for now

NSeydoux committed:
fixup! Move crypto polyfilling to the polyfill module

NSeydoux committed:
Fix failing test

The core module mocking was conflicting with the oidc-ext one.

NSeydoux committed:
Clear open handles on core package

NSeydoux committed:
Lint

NSeydoux committed:
Centralize all jest config in root

Using the projects array and --selectProject CLI option, we can get rig of all subprojects jest configs and manage it all from one root file.

ThisIsMissEm commented on jest.config.ts:
```diff
@@ -27,15 +28,48 @@ export default {
+  }, {
+    ...baseConfig,
+    displayName: "oidc",
+    roots: ["<rootDir>/packages/oidc"],
```
For clarity, would it be worth renaming this folder to `oidc-browser` (not changing the package name, just the folder name)

NSeydoux replied:
No strong opinion, but no harm in being more explicit: 1a94ef691cf16b8a33eb8530c406e7e7584979b0

NSeydoux replied:
Actually this breaks sonar integration, which would create a bit of delay to sort out where this config is. I'd like this PR to get merged now, as it has already lingered quite a while, so let's do the renaming later when I have figured out what needs to be adjusted on sonar.

ThisIsMissEm replied:
It should just be updating here: https://github.com/inrupt/solid-client-authn-js/blob/main/.github/workflows/ci.yml#L63

ThisIsMissEm replied:
But I am of course okay doing it separatey

NSeydoux replied:
Ah cool, it was late yesterday when I was looking into it so I didn't want to investigate, but if that's just that's good.

ThisIsMissEm commented on packages/browser/package.json:
```diff
@@ -22,7 +22,7 @@
     "lint:prettier": "prettier \"src/*.{ts,tsx,js,jsx,css}\" \"**/*.{md,mdx,yml}\"",
     "licenses:check": "license-checker --production --out license.csv --failOn \"AGPL-1.0-only; AGPL-1.0-or-later; AGPL-3.0-only; AGPL-3.0-or-later; Beerware; CC-BY-NC-1.0; CC-BY-NC-2.0; CC-BY-NC-2.5; CC-BY-NC-3.0; CC-BY-NC-4.0; CC-BY-NC-ND-1.0; CC-BY-NC-ND-2.0; CC-BY-NC-ND-2.5; CC-BY-NC-ND-3.0; CC-BY-NC-ND-4.0; CC-BY-NC-SA-1.0; CC-BY-NC-SA-2.0; CC-BY-NC-SA-2.5; CC-BY-NC-SA-3.0; CC-BY-NC-SA-4.0;  ...
-    "test": "OPENSSL_CONF=/dev/null jest --coverage --verbose",
+    "test": "OPENSSL_CONF=/dev/null jest --coverage --verbose --selectProjects browser --config ../../jest.config.ts",
```
Can this be done with a setupFiles setting that mutates process.env? (That is, before the test code is loaded)

NSeydoux replied:
Interestingly, it can be done in jest's config file, but not as part of the environment setup for some reason: d18fbcb4cf8662ae90410c16b7674b0931e8dc39.

NSeydoux committed:
Update jest.config.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Move environment setup to jest config

Having it in the jest setup file doesn't work.

NSeydoux committed:
Rename packages/oidc into oidc-browser

NSeydoux committed:
Revert "Rename packages/oidc into oidc-browser"

This reverts commit 1a94ef691cf16b8a33eb8530c406e7e7584979b0.
Some configs, such as sonar scan, depend on the oidc folder name, and I really want this to get merged so we can rename this another time.

NSeydoux committed:
Revert "Revert "Rename packages/oidc into oidc-browser""

This reverts commit 02084c57003dd74f41c45189d39bbf378a025fd8.

NSeydoux committed:
fixup! Rename packages/oidc into oidc-browser

ThisIsMissEm commented on jest.config.ts:
```diff
@@ -0,0 +1,78 @@
+const baseConfig: JestConfigWithTsJest = {
+  roots: ["<rootDir>"],
+  clearMocks: true,
+  testMatch: ["**/(__tests__|src)/**/?(*.)+(spec|test).+(ts|tsx|js)"],
```
We should really standardise around either `.test.ts` or `.spec.ts`, but not both. Maybe we can already change this to: 
```suggestion
  testMatch: ["src/**/(.+).spec.ts"],
```

Elsewhere we use: 
```
testRegex: "/src/.*\\.test\\.ts$",
```

This is definitely a future problem to tackle though. But lets standardise across all our projects.

NSeydoux replied:
This repo should consistently be using `.spec.ts`, and I have no problem changing to `.test.ts`, just in another PR :)

ThisIsMissEm commented on packages/browser/src/login/oidc/incomingRedirectHandler/AuthCodeRedirectHandler.ts:
```diff
@@ -149,12 +149,16 @@ export class AuthCodeRedirectHandler implements IIncomingRedirectHandler {
-      expiresIn: tokens.expiresIn,
-    });
+    const authFetch = await buildAuthenticatedFetch(
+      window.fetch,
```
should this be `window.fetch` or should it be `import { fetch } from "cross-fetch";`? (I mean, yeah, arguably we are in the browser, so we have `window.fetch` but I think maybe we want to standardise around the import, especially with our Node.js 18+ support work.

NSeydoux replied:
Fixed in 535276759bd27391d0c211ccca72d2531de128f8, but there's a lot of that remaining :)

ThisIsMissEm commented on packages/core/package.json:
```diff
@@ -30,29 +31,23 @@
     "clean": "npm run clean-module",
     "clean-module": "rimraf ./dist",
-    "build": "tsc -p tsconfig.json",
+    "build": "rollup --config rollup.config.mjs",
```
ooohh! aaahh! You've done rollup here too! Nice!!

NSeydoux replied:
It's just for the core though, I think the browser relies on WebPack a bit more so that will be more of a pickle. Here is allows to output ESM as well as CJS (which I don't think was actually needed in retrospective ^^)

ThisIsMissEm commented on packages/core/package.json:
```diff
@@ -30,29 +31,23 @@
-  },
   "dependencies": {
-    "@inrupt/solid-common-vocab": "^1.0.0",
     "@types/lodash.clonedeep": "^4.5.6",
     "@types/uuid": "^8.3.0",
```
These need to be moved to `devDependencies`

ThisIsMissEm commented on packages/core/rollup.config.mjs:
```diff
@@ -0,0 +1,62 @@
+      format: "esm",
+    },
+  ],
+  plugins: [
+    typescript({
+      // Use our own version of TypeScript, rather than the one bundled with the plugin:
+      typescript: require("typescript"),
+      tsconfigOverride: {
+        compilerOptions: {
+          module: "esnext",
+        },
+      },
+    })
+  ],
+  // The following option is useful because symlinks are used in monorepos
+  preserveSymlinks: true,
```
Let's move this into a `common` const, as it's used in both exports (index & mocks)

ThisIsMissEm commented on packages/core/src/authenticatedFetch/fetchFactory.spec.ts:
```diff
@@ -98,6 +98,17 @@ const mockFetch = (response: Response, url: string) => {
 };
 
 describe("buildAuthenticatedFetch", () => {
+  const spyTimeout = jest.spyOn(global, "setTimeout");
+
+  afterEach(() => {
+    // Clear the latest timeout to avoid dangling open handles.
+    // FIXME: Should just use fake timers, but that chokes on recursive calls.
+    const handle = spyTimeout.mock.results[spyTimeout.mock.results.length - 1];
+    if (handle !== undefined) {
+      (handle.value as ReturnType<typeof setTimeout>).unref();
+    }
+  });
```
Can this be changed with the usage of mock timers? https://jestjs.io/docs/timer-mocks — err, totally blanked on the comment; Let's open a Jira ticket to fix this.

ThisIsMissEm replied:
I wonder if the issue you faces was due to https://jestjs.io/docs/timer-mocks#run-pending-timers

i.e., `jest.runOnlyPendingTimers();` vs `jest.runAllTimers();`

NSeydoux replied:
Unfortunately no it didn't appear to be the issue. Basically it looks like Jest handles the timers being set by `setTimeout`, but not the ones set in the callback passed to `setTimeout`, resulting in the second round of timers not to be mocked.

ThisIsMissEm commented on packages/core/src/authenticatedFetch/fetchFactory.spec.ts:
```diff
@@ -314,7 +330,9 @@ describe("buildAuthenticatedFetch", () => {
 
   // jest.useFakeTimers();
   function sleep(ms: number) {
-    return new Promise((resolve) => setTimeout(resolve, ms));
+    return new Promise((resolve) => {
+      setTimeout(resolve, ms);
+    });
```
When we use fake timers, we'll need to change these calls to `jest.advanceTimersByTime(ms);`

ThisIsMissEm commented on packages/core/src/login/oidc/__mocks__/IncomingRedirectHandler.ts:
```diff
@@ -22,23 +22,17 @@
+  Promise.resolve({
+    sessionId: "global",
+    isLoggedIn: true,
+    webId: "https://pod.com/profile/card#me",
```
```suggestion
    webId: "https://webid.test/profile/card#me",
```

We shouldn't use real URLs in tests, `pod.com` is potentially real

NSeydoux replied:
I'll add a tech debt ticket to do this all at once, it applies across all codebase

ThisIsMissEm commented on packages/oidc-browser/package.json:
```diff
@@ -20,20 +20,21 @@
   "dependencies": {
     "@inrupt/oidc-client": "^1.11.6",
     "@inrupt/solid-client-authn-core": "^1.12.2",
     "@types/jest": "^27.0.3",
     "@types/uuid": "^8.3.0",
```
`@types` need to be moved to `devDependencies`

ThisIsMissEm commented on packages/oidc-browser/src/refresh/refreshGrant.spec.ts:
```diff
@@ -48,6 +48,15 @@ jest.mock("@inrupt/solid-client-authn-core", () => {
+//     getWebidFromTokenPayload: jest.fn<() => Promise<string>>().mockResolvedValue("https://my.webid/")
+//   };
+// });
+
```
Can this be deleted entirely?

ThisIsMissEm reviewed (changes requested):
Looks good! Just a few more things left to resolve — Feel free to comment on them if you split them into separate work tickets.

NSeydoux commented on packages/browser/src/Session.ts:
```diff
@@ -159,7 +159,7 @@ export class Session extends EventEmitter {
   constructor(
     sessionOptions: Partial<ISessionOptions> = {},
-    sessionId?: string
+    sessionId: string | undefined = undefined
```
Deprecating the session id being nullable on construction? I don't think so.

ThisIsMissEm replied:
I thought it was moved into sessionOptions

NSeydoux commented on packages/browser/src/login/oidc/incomingRedirectHandler/AuthCodeRedirectHandler.spec.ts:
```diff
@@ -143,37 +146,22 @@ jest.mock("@inrupt/oidc-client-ext");
+    "@inrupt/oidc-client-ext"
   );
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  mockedOidcClient.decodeJwt = (jest.fn() as any).mockResolvedValue(jwt);
```
Yeah I'm not sure how that got there in the first place ^

NSeydoux committed:
Update jest.config.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Update jest.config.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Update jest.config.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Fix test match regex

NSeydoux commented on jest.config.ts:
```diff
@@ -0,0 +1,78 @@
+  collectCoverageFrom: [
+    "<rootDir>/src/**/*.ts",
+  ],
+  coveragePathIgnorePatterns: [
+    "node_modules/",
+    "dist/",
+    ".*.spec.ts"
+  ],
```
`dist` appears to be required there, but not node_modules.

NSeydoux committed:
Remove node_modules from ignored coverage

redundant

NSeydoux committed:
Apply feedback

NSeydoux committed:
Removed some fetch mock inconsistencies

NSeydoux committed:
Reduce redundancy in rollup config

NSeydoux committed:
Fix dev deps being prod deps

NSeydoux committed:
Move prod deps to dev deps

NSeydoux committed:
Remove commented out code

NSeydoux committed:
fixup! Apply feedback

NSeydoux committed:
Lint

NSeydoux committed:
Using cross-fetch breaks browser module

NSeydoux committed:
fixup! Using cross-fetch breaks browser module

NSeydoux committed:
fixup! fixup! Using cross-fetch breaks browser module

NSeydoux committed:
Hoist all jest interaction to top level

Instead of relying on Lerna to run tests across the repos, just use Jest picking a specific project or running them all.
